xclim/subset.py

```diff
@@ -125,11 +133,188 @@ def func_checker(*args, **kwargs):
+     the `rioxarray` library imported. Attempting to use this function with pure xarray objects will raise exceptions.
+
+    Examples
+    --------
+    >>> from xclim import subset
+    >>> import rioxarray
+    >>> import xarray as xr
+    >>> ds = xarray.open_dataset('pr.day.nc')
+    Subset lat lon and years
+    >>> prSub = subset.subset_shape(ds.pr, shape="/path/to/polygon.shp", start_yr='1990', end_yr='1999')
+    Subset data array lat, lon and single year
+    >>> prSub = subset.subset_shape(ds.pr, shape="/path/to/polygon.shp", start_yr='1990', end_yr='1990')
+    Subset data array single year keep entire lon, lat grid
+    >>> prSub = subset.subset_bbox(ds.pr, start_yr='1990', end_yr='1990') # one year only entire grid
+    Subset multiple variables in a single dataset
+    >>> ds = xarray.open_mfdataset(['pr.day.nc','tas.day.nc'])
+    >>> dsSub = subset.subset_bbox(ds, shape="/path/to/polygon.shp", start_yr='1990', end_yr='1999')
+     # Subset with year-month precision - Example subset 1990-03-01 to 1999-08-31 inclusively
+    >>> prSub = subset.subset_time(ds.pr, shape="/path/to/polygon.shp", start_date='1990-03', end_date='1999-08')
+    # Subset with specific start_dates and end_dates
+    >>> prSub = \
+            subset.subset_time(ds.pr, shape="/path/to/polygon.shp", start_date='1990-03-13', end_date='1990-08-17')
+    """
+    try:
+        if da.rio.crs is None:
+            if np.any(da.lon < -180) or np.any(da.lon > 360):
+                raise rasterio.crs.CRSError("Set")
+            if np.any(da.lon > 180):
+                lon_attrs = da.lon.attrs.copy()
+                fix_lon = da.lon.values
+                fix_lon[fix_lon > 180] = fix_lon[fix_lon > 180] - 360
+                if "lon_bnds" in da.data_vars:
+                    fix_lon_bnds = da.lon_bnds.values
+                    fix_lon_bnds[fix_lon_bnds > 180] = (
+                        fix_lon_bnds[fix_lon_bnds > 180] - 360
+                    )
+                da = da.assign_coords(lon=fix_lon)
+                da = da.sortby("lon")
+                da.lon.attrs = lon_attrs
+            if da_crs is None:
+                crs = rasterio.crs.CRS.from_epsg(4326)
+            else:
+                crs = rasterio.crs.CRS.from_user_input(da_crs)
+        else:
+            crs = da.rio.crs
+        if shape_crs != crs:
+            raise rasterio.crs.CRSError("Shape and Raster CRS are not the same.")
+    except Exception as e:
+        logging.exception(e)
+        raise
+
+    ds_out = xarray.Dataset(data_vars=None, attrs=da.attrs)
+    for v in da.data_vars:
+        if "lon" in da[v].dims and "lat" in da[v].dims:
+            dss = da[v]
+            dss.rio.set_spatial_dims(x_dim="lon", y_dim="lat", inplace=True)
+            dss.rio.write_crs(crs, inplace=True)
+            ds_out[v] = dss.rio.clip(
+                geometry, crs=dss.rio.crs, all_touched=True, drop=True, invert=False
```

Comment:
Is it possible to get the weights as well ? That is, the fraction of each grid cell's area included in the subset.

Reply:
If it's not provided out of the box by rioxarray, I suggest we open a ticket for this.

Reply:
Zonal statistics is something in rioxarray but it requires a bit of work on both xclim and rioxarray. I can look more into this in another PR if we want to have regional statistics reporting.

Reply:
To be clear, we'd be interested in having, for each cell within the subset, the percent area of the total region area. This would account for the both partial intersection of a grid with the polygon, and differences in individual grid cell areas. Ok, let's open a new PR once this one's done.